File: src/compiler/lexer/token/OperatorToken.java
package compiler.lexer.token;

import compiler.parser.TokenEvaluator;
import compiler.parser.visitors.TokenVisitor;

public abstract class OperatorToken extends Token {

  private OperatorToken(String str, int UUID) {
    super(str, UUID);
  }

  public static class LessThan extends OperatorToken {
    public LessThan() {
      super("<", 31);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class GreaterThan extends OperatorToken {
    public GreaterThan() {
      super(">", 32);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Asterisk extends OperatorToken {
    public Asterisk() {
      super("*", 41);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Object accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Equal extends OperatorToken {
    public Equal() {
      super("=", 45);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Void accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Minus extends OperatorToken {
    public Minus() {
      super("-", 46);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Number accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Plus extends OperatorToken {
    public Plus() {
      super("+", 47);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Number accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Ampersand extends OperatorToken {
    public Ampersand() {
      super("&", 49);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Object accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class Arrow extends OperatorToken {
    public Arrow() {
      super("->", 51);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Void accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class EqualEqual extends OperatorToken {
    public EqualEqual() {
      super("==", 52);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class NotEqual extends OperatorToken {
    public NotEqual() {
      super("!=", 53);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class LessThanOrEqual extends OperatorToken {
    public LessThanOrEqual() {
      super("<=", 54);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class GreaterThanOrEqual extends OperatorToken {
    public GreaterThanOrEqual() {
      super(">=", 55);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Boolean accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class BitShiftLeft extends OperatorToken {
    public BitShiftLeft() {
      super("<<", 56);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Number accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }

  public static class BitShiftRight extends OperatorToken {
    public BitShiftRight() {
      super(">>", 57);
    }

    @Override
    public <T> T accept(TokenVisitor<T> visitor) {
      return visitor.visit(this);
    }


    @Override
    public Number accept(TokenEvaluator visitor) throws Exception {
      return visitor.visit(this);
    }
  }
}

File: src/compiler/lexer/token/OperatorToken.kt
package compiler.lexer.token

import compiler.parser.TokenEvaluator
import compiler.parser.visitors.TokenVisitor

abstract class OperatorToken private constructor(str: String, UUID: Int) : Token(str, UUID) {
    class LessThan : OperatorToken("<", 31) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class GreaterThan : OperatorToken(">", 32) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class Asterisk : OperatorToken("*", 41) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Any {
            return visitor.visit(this)
        }
    }

    class Equal : OperatorToken("=", 45) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator) {
            return visitor.visit(this)
        }
    }

    class Minus : OperatorToken("-", 46) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }

    class Plus : OperatorToken("+", 47) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }

    class Ampersand : OperatorToken("&", 49) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Any {
            return visitor.visit(this)
        }
    }

    class Arrow : OperatorToken("->", 51) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator) {
            return visitor.visit(this)
        }
    }

    class EqualEqual : OperatorToken("==", 52) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class NotEqual : OperatorToken("!=", 53) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class LessThanOrEqual : OperatorToken("<=", 54) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class GreaterThanOrEqual : OperatorToken(">=", 55) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Boolean {
            return visitor.visit(this)
        }
    }

    class BitShiftLeft : OperatorToken("<<", 56) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }

    class BitShiftRight : OperatorToken(">>", 57) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }

    class Caret : OperatorToken("^", 42) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }

    class ForwardSlash : OperatorToken("/", 48) {
        override fun <T> accept(visitor: TokenVisitor<T>): T {
            return visitor.visit(this)
        }

        @Throws(Exception::class)
        override fun accept(visitor: TokenEvaluator): Number {
            return visitor.visit(this)
        }
    }
}
